File: go_src/internal/model/manifest.go
// Manifest is the routing index the dispatcher picks on (manifest.json). The summaries are the only
// thing routing sees; the host reads this to know what KB entries exist and where they live. Port of
// src.bak/Model/Manifest.cs.
package model

import (
	"encoding/json"
	"fmt"
	"os"
	"sort"
	"strings"
)

// Entry is one routable KB document's metadata. Field order matches the C# writer (id, title, path,
// summary, doc_type, group, keywords) so regenerated manifests are byte-stable.
type Entry struct {
	ID       string   `json:"id"`
	Title    string   `json:"title"`
	Path     string   `json:"path"`
	Summary  string   `json:"summary"`
	DocType  string   `json:"doc_type"`
	Group    string   `json:"group"` // sub-folder under the routable layer, e.g. "dotnet" or "creational"
	Keywords []string `json:"keywords"`
}

// Manifest is a named index of entries.
type Manifest struct {
	Name        string
	Description string
	Entries     []Entry
}

type manifestJSON struct {
	Name        string  `json:"name"`
	Description string  `json:"description"`
	Entries     []Entry `json:"entries"`
}

// LoadManifest reads and parses a manifest.json.
func LoadManifest(path string) (*Manifest, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("reading %s: %v", path, err)
	}
	var raw manifestJSON
	if err := json.Unmarshal(data, &raw); err != nil {
		return nil, fmt.Errorf("parsing %s: %v", path, err)
	}
	m := &Manifest{Name: raw.Name, Description: raw.Description}
	for _, e := range raw.Entries {
		if e.Keywords == nil {
			e.Keywords = []string{}
		}
		m.Entries = append(m.Entries, e)
	}
	return m, nil
}

// GetEntry returns the entry with the given id, or nil.
func (m *Manifest) GetEntry(id string) *Entry {
	for i := range m.Entries {
		if m.Entries[i].ID == id {
			return &m.Entries[i]
		}
	}
	return nil
}

// Groups returns the distinct, sorted sub-folder groups present in the index.
func (m *Manifest) Groups() []string {
	var seen []string
	known := map[string]bool{}
	for _, e := range m.Entries {
		if e.Group != "" && !known[e.Group] {
			known[e.Group] = true
			seen = append(seen, e.Group)
		}
	}
	sort.Slice(seen, func(i, j int) bool { return strings.ToLower(seen[i]) < strings.ToLower(seen[j]) })
	return seen
}

// ByGroup returns the entries in the given group (case-insensitive).
func (m *Manifest) ByGroup(group string) []Entry {
	var o []Entry
	for _, e := range m.Entries {
		if strings.EqualFold(e.Group, group) {
			o = append(o, e)
		}
	}
	return o
}

// ByDocType returns the entries of the given doc_type (case-insensitive).
func (m *Manifest) ByDocType(docType string) []Entry {
	var o []Entry
	for _, e := range m.Entries {
		if strings.EqualFold(e.DocType, docType) {
			o = append(o, e)
		}
	}
	return o
}

// Catalog renders a compact "id [group]: summary" listing, optionally filtered by group and/or
// doc_type. An empty filter means "no filter on that field". This is the catalog the model enumerates.
func (m *Manifest) Catalog(group, docType string) string {
	var sb strings.Builder
	for _, e := range m.Entries {
		if group != "" && !strings.EqualFold(e.Group, group) {
			continue
		}
		if docType != "" && !strings.EqualFold(e.DocType, docType) {
			continue
		}
		g := ""
		if e.Group != "" {
			g = " [" + e.Group + "]"
		}
		sb.WriteString("- " + e.ID + g + ": " + e.Summary + "\n")
	}
	return strings.TrimRight(sb.String(), "\n")
}

File: csharp_src/Model/Manifest.cs
// Manifest loader (manifest.json): the routing index the dispatcher picks on. Port of
// manifest.rs. The summaries are the only thing routing sees; the host reads this to know what
// KB entries exist and where they live.

using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace Icm
{
    internal class Entry
    {
        public string Id = "";
        public string Title = "";
        public string Path = "";
        public string Summary = "";
        public string DocType = "";
        public string Group = "";   // sub-folder under the routable layer, e.g. "dotnet" or "creational"
        public List<string> Keywords = new List<string>();
    }

    internal class Manifest
    {
        public string Name = "";
        public string Description = "";
        public List<Entry> Entries = new List<Entry>();

        public static Manifest Load(string path)
        {
            string text;
            try { text = File.ReadAllText(path); }
            catch (Exception e) { throw new IcmError("reading " + path + ": " + e.Message); }

            Dictionary<string, object> root;
            try { root = Json.AsObject(Json.Parse(text)); }
            catch (Exception e) { throw new IcmError("parsing " + path + ": " + e.Message); }
            if (root == null) throw new IcmError("parsing " + path + ": not a JSON object");

            var m = new Manifest();
            m.Name = Json.GetStringOr(root, "name", "");
            m.Description = Json.GetStringOr(root, "description", "");
            foreach (object e in Json.GetArr(root, "entries"))
            {
                var eo = e as Dictionary<string, object>;
                if (eo == null) continue;
                var entry = new Entry();
                entry.Id = Json.GetStringOr(eo, "id", "");
                entry.Title = Json.GetStringOr(eo, "title", "");
                entry.Path = Json.GetStringOr(eo, "path", "");
                entry.Summary = Json.GetStringOr(eo, "summary", "");
                entry.DocType = Json.GetStringOr(eo, "doc_type", "");
                entry.Group = Json.GetStringOr(eo, "group", "");
                foreach (object kw in Json.GetArr(eo, "keywords")) if (kw != null) entry.Keywords.Add(kw.ToString());
                m.Entries.Add(entry);
            }
            return m;
        }

        public Entry GetEntry(string id)
        {
            foreach (Entry e in Entries) if (e.Id == id) return e;
            return null;
        }

        // --- enumeration: the deterministic, host-owned views the model proposes against ---

        // Distinct, sorted sub-folder groups present in the index.
        public List<string> Groups()
        {
            var seen = new List<string>();
            foreach (Entry e in Entries)
                if (e.Group.Length > 0 && !seen.Contains(e.Group)) seen.Add(e.Group);
            seen.Sort(StringComparer.OrdinalIgnoreCase);
            return seen;
        }

        public List<Entry> ByGroup(string group)
        {
            var o = new List<Entry>();
            foreach (Entry e in Entries) if (string.Equals(e.Group, group, StringComparison.OrdinalIgnoreCase)) o.Add(e);
            return o;
        }

        public List<Entry> ByDocType(string docType)
        {
            var o = new List<Entry>();
            foreach (Entry e in Entries) if (string.Equals(e.DocType, docType, StringComparison.OrdinalIgnoreCase)) o.Add(e);
            return o;
        }

        // A compact "id [group]: summary" listing, optionally filtered by group and/or doc_type.
        // Null/empty filter means "no filter on that field". This is the catalog the model enumerates.
        public string Catalog(string group, string docType)
        {
            var sb = new StringBuilder();
            foreach (Entry e in Entries)
            {
                if (!string.IsNullOrEmpty(group) && !string.Equals(e.Group, group, StringComparison.OrdinalIgnoreCase)) continue;
                if (!string.IsNullOrEmpty(docType) && !string.Equals(e.DocType, docType, StringComparison.OrdinalIgnoreCase)) continue;
                string g = e.Group.Length > 0 ? " [" + e.Group + "]" : "";
                sb.Append("- " + e.Id + g + ": " + e.Summary + "\n");
            }
            return sb.ToString().TrimEnd();
        }
    }
}
